server.py

```diff
@@ -0,0 +1,169 @@
+                mask_points = list(map(lambda x: x.tolist(), mask_points_nd))
+                objects[i-1]["mask"] = mask_points
+                objects[i-1]["type"] = "annotate"
+
+            # Returns an array of objects with updated masks
+            sio.emit("labelPropagationReturn", objects)
+        
+        @sio.on("offline_save_rgb_seg")
+        def offline_save_rgb_seg(sid, postData): 
+
+            # Get rgb image
+            rgb_path = os.path.join(postData["filepath"], "rgb")
+            num_zeros = 5 - len(str(postData["frameId"]))
+            file_num = "".join(["0" for _ in range(num_zeros)]) + str(postData["frameId"])
+            rgb_filename = os.path.join(rgb_path, file_num + ".jpg")
+            rgb = cv2.imread(rgb_filename)
+            height, width, _ = rgb.shape
+
+            # Convert mask points to mask maps then combine them
+            categories = postData["categories"]
+            display_map = np.zeros((height, width)).astype(int) # 2 separate chair masks will be same color here
+            for o in postData["objects"]: 
+                poly = Polygons(o["mask"])
+                bitmap = poly.mask(height, width)
+                index = categories.index(o["label"])
+                for i in range(height): 
+                    for j in range(width): 
+                        if bitmap[i][j]: 
+                            display_map[i][j] = index
+
+            # Save annotation data to disk for retraining
+            Path("annotation_data/seg").mkdir(parents=True, exist_ok=True)
+            Path("annotation_data/rgb").mkdir(parents=True, exist_ok=True)
+            np.save("annotation_data/seg/{:05d}.npy".format(postData["outputId"]), display_map)
+            im = Image.fromarray(rgb)
+            im.save("annotation_data/rgb/{:05d}.jpg".format(postData["outputId"]))
+
+            if postData["finalFrame"]: 
+                LP.save_annotations(postData["categories"])
+
+        @sio.on("save_categories_properties")
+        def save_categories_properties(sid, categories, properties): 
+            LP.save_categories_properties(categories, properties)
+
+        @sio.on("retrain_detector")
+        def retrain_detector(sid, settings={}): 
+            inference_json = LP.retrain_detector(settings)
+            sio.emit("annotationRetrain", inference_json)
+        
+        # Send first frame in folder
+        @sio.on("start_offline_dashboard")
+        def start_offline_dashboard(sid, filepath): 
+            num_files = len(os.listdir(os.path.join(filepath, "rgb")))
+            sio.emit("handleMaxFrames", num_files - 1) # Minus 1 because filenames start at 00000
+
+        # Send first frame in folder
+        @sio.on("get_offline_frame")
+        def get_offline_frame(sid, data): 
+            rgb, depth = LP.get_offline_frame(data)
+            sio.emit("rgb", rgb)
+            sio.emit("depth", depth)
+
+
+if __name__ == "__main__":
+    
+    base_path = os.path.dirname(__file__)
```

Comment:
how is this server kicked off while the dashboard is running?

Reply:
Server.py is the backend for the offline dashboard and is the agent-free replacement for locobot_agent.py